A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,258 @@
+
+### Subchannel Wrapper
+
+The subchannel wrappers created by the `outlier_detection` LB policy will hold a reference to its map entry in the LB policy, if that map entry exists.
+
+The subchannel wrapper will track the latest state update from the underlying subchannel. By default, it will simply pass those updates along. The wrapper will have the following methods:
+
+ - `eject()`: The wrapper will report a state update with the `TRANSIENT_FAILURE` state, and will stop passing along updates from the underlying subchannel.
+ - `uneject()`: The wrapper will report a state update with the latest update from the underlying subchannel, and resume passing along updates from the underlying subchannel.
+
+### Success Rate Algorithm
+
+ 1. If the number of addresses with request volume of at least `success_rate_ejection.request_volume` is less than `success_rate_ejection.minimum_hosts`, stop.
+ 2. Calculate the mean and standard deviation of the fractions of successful requests among addresses with total request volume of at least `success_rate_ejection.request_volume`.
+ 3. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `success_rate_ejection.request_volume`, continue to the next address.
+    3. If the address's success rate is less than `(mean - stdev * (success_rate_ejection.stdev_factor / 1000))`, then choose a random integer in `[0, 100)`. If that number is less than `success_rate_ejection.enforcement_percentage`, eject that address.
+
+### Failure Percentage Algorithm
+
+ 1. If the number of addresses is less than `failure_percentage_ejection.minimum_hosts`, stop.
+ 2. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `failure_percentage_ejection.request_volume`, continue to the next address.
+    3. If the address's failure percentage is greater than `failure_percentage_ejection.threshold`, then choose a random integer in `[0, 100)`. If that number is less than `failiure_percentage_ejection.enforcement_percentage`, eject that address.
+
+### xDS Integration
+
+If the `Cluster` message for an underlying cluster has the `outlier_detection` field set, the `xds_cluster_resolver` will generate an `outlier_detection` LB policy config from it for each priority as the top-level LB policy for that priority. In that `outlier_detection` field message, if the `enforcement_percentage` field is set to 0, the config `success_rate_ejection` field will be `null` and all `success_rate_*` fields will be ignored. If the `enforcing_failure_percent` field is set to 0 or `null`, the config `failure_percent_ejection` field will be `null` and all `failure_percent_*` fields will be ignored. Then the message fields will be mapped to config fields as follows:
```

Comment:
Since `outlier_detection` is a field in the `Cluster` message for the underlying cluster, does that mean that this information should be included in the `DiscoveryMechanism` message in the `xds_cluster_resolver` LB policy config? Also, would it be reasonable to just add an `outlier_detection` field in that config with a message type that is almost identical to the actual `outlier_detection` LB policy config except for the `child_policy` field?

Reply:
Yes, you're correct, it will need to be inside of the `DiscoveryMechanism` message.

For the proto message, I see a few options:
1. Duplicate the `outlier_detection` LB policy config except for the `child_policy` field.  The down-side of this is that it would require essentially maintaining two copies going forward, although this message probably won't change very frequently, so that may not be a big deal.
2. Change the `outlier_detection` LB policy config proto such that all of the fields except `child_policy` are in a separate message, so that we can reuse that message in both places.  The down-side of this is that it adds another level of nesting in the common case that users actually have to deal with in the non-xDS case.
3. Keep the `outlier_detection` LB policy config as-is and re-use it completely, just adding a comment that the `child_policy` field will be ignored in the `xds_cluster_resolver` config.  This is a little ugly in terms of proto semantics, but since our code actually parses these configs in JSON form instead of proto form, that ugliess won't make any practical difference.

My inclination would be to go with option 3.

Note that in the implementation, we will want to have one piece of code to parse all of the common fields, rather than duplicating that functionality in both policies.

Reply:
I made an update addressing this comment and the other one. How does it look now?